File: internal/gitops/message.go
package gitops

import (
	"bytes"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"text/template"
	"time"
)

// DefaultTemplate is the shipped commit-message template
// (docs/06-git-sync.md section 3.3).
const DefaultTemplate = `pmngr: update {{.ItemID}} "{{.Title}}"`

// SubjectLimit is the length a subject line is truncated to. The full title
// stays in the body, so nothing is lost.
const SubjectLimit = 72

// ToolName is the `Tool:` trailer prefix. The caller appends its version.
const ToolName = "gintrack"

// Action is what a write did, as the `{{.Action}}` placeholder reports it.
type Action string

// The actions of docs/06 section 3.3.
const (
	ActionCreate  Action = "create"
	ActionUpdate  Action = "update"
	ActionDelete  Action = "delete"
	ActionMove    Action = "move"
	ActionComment Action = "comment"
)

// String renders the action.
func (a Action) String() string {
	if a == "" {
		return string(ActionUpdate)
	}
	return string(a)
}

// Fields is the template context of docs/06 section 3.3 and docs/07 section
// 6.4. Every field is a plain string so that a template can never fail on a nil
// dereference; Count is the only number and is 1 for a single-item write.
type Fields struct {
	ItemID     string `json:"itemId,omitempty"`
	Title      string `json:"title,omitempty"`
	Type       string `json:"type,omitempty"`
	Status     string `json:"status,omitempty"`
	PrevStatus string `json:"prevStatus,omitempty"`
	ProjectKey string `json:"projectKey,omitempty"`
	Board      string `json:"board,omitempty"`
	Action     Action `json:"action,omitempty"`
	Count      int    `json:"count,omitempty"`
	User       string `json:"user,omitempty"`
	Date       string `json:"date,omitempty"`
	// Tool is the `Tool:` trailer value, for example
	// "gintrack 0.4.1 (companion)". It is not a template placeholder.
	Tool string `json:"tool,omitempty"`
	// Agent is the `Agent:` trailer written when the change came from the MCP
	// server (docs/07 section 6.4). Empty for a human edit.
	Agent string `json:"agent,omitempty"`
	// Extra carries trailers a caller wants beyond the standard set.
	Extra map[string]string `json:"extra,omitempty"`
}

// bulk reports whether the fields describe several items at once, which is what
// a bulk status change or a `updateMany` produces.
func (f Fields) bulk() bool { return f.Count > 1 && f.ItemID == "" }

// Template is a parsed, reusable commit-message template.
type Template struct {
	text string
	tpl  *template.Template
}

// Text returns the template source, so the settings surface can echo back what
// is configured.
func (t *Template) Text() string {
	if t == nil {
		return DefaultTemplate
	}
	return t.text
}

// ParseTemplate compiles a commit-message template. An empty text compiles the
// shipped default.
//
// Two spellings are accepted for the same data, because both are documented:
// the Go text/template field form (`{{.ItemID}}`, docs/06 section 3.3) and the
// short lowercase form (`{{action}} {{id}}: {{title}}`, story GIT-US-0020),
// which is implemented as niladic template functions over the same fields.
func ParseTemplate(text string) (*Template, error) {
	if strings.TrimSpace(text) == "" {
		text = DefaultTemplate
	}
	// The functions are re-bound per render; parsing only needs their names to
	// exist, so a placeholder map with the right arity is enough here.
	tpl, err := template.New("commit").Funcs(aliasFuncs(Fields{})).Parse(text)
	if err != nil {
		return nil, wrap("template", CodeTemplateInvalid, err,
			"the commit message template does not parse")
	}
	return &Template{text: text, tpl: tpl}, nil
}

// MustParseTemplate is ParseTemplate for a template known to be valid, such as
// the shipped default. It panics on a broken one, which can only be a bug.
func MustParseTemplate(text string) *Template {
	t, err := ParseTemplate(text)
	if err != nil {
		panic(err)
	}
	return t
}

// aliasFuncs binds the short lowercase placeholders to the fields.
func aliasFuncs(f Fields) template.FuncMap {
	return template.FuncMap{
		"id":         func() string { return f.ItemID },
		"title":      func() string { return f.Title },
		"type":       func() string { return f.Type },
		"status":     func() string { return f.Status },
		"prevStatus": func() string { return f.PrevStatus },
		"project":    func() string { return f.ProjectKey },
		"board":      func() string { return f.Board },
		"action":     func() string { return f.Action.String() },
		"count":      func() int { return f.count() },
		"user":       func() string { return f.User },
		"date":       func() string { return f.Date },
	}
}

// count is Count with the 1 a single-item write leaves implicit.
func (f Fields) count() int {
	if f.Count < 1 {
		return 1
	}
	return f.Count
}

// Render produces the commit message: the subject from the template and the
// body from the machine-readable trailers of docs/06 section 3.3.
//
// The subject never contains a newline and is truncated to SubjectLimit; a
// truncated title is repeated in full in the body, so the information survives.
func (t *Template) Render(f Fields) (Message, error) {
	if t == nil {
		t = MustParseTemplate(DefaultTemplate)
	}
	f = f.normalise()

	subject, err := t.subject(f)
	if err != nil {
		return Message{}, err
	}

	body := trailers(f)
	if len(subject) > SubjectLimit {
		body = append([]string{"Title: " + f.Title, ""}, body...)
		subject = truncate(subject, SubjectLimit)
	}
	return Message{Subject: subject, Body: strings.Join(body, "\n")}, nil
}

// subject renders the template, or the built-in bulk subject when the fields
// describe several items and therefore have no single id to interpolate.
func (t *Template) subject(f Fields) (string, error) {
	if f.bulk() {
		return fmt.Sprintf("pmngr: %s %d items", f.Action.String(), f.count()), nil
	}
	var buf bytes.Buffer
	tpl, err := t.tpl.Clone()
	if err != nil {
		return "", wrap("template", CodeTemplateInvalid, err, "the commit message template could not be prepared")
	}
	if err := tpl.Funcs(aliasFuncs(f)).Execute(&buf, f); err != nil {
		return "", wrap("template", CodeTemplateInvalid, err,
			"the commit message template does not render")
	}
	subject := collapse(buf.String())
	if subject == "" {
		return "", failf("template", CodeTemplateInvalid,
			"the commit message template rendered an empty subject")
	}
	return subject, nil
}

// normalise fills the defaults a template may rely on.
func (f Fields) normalise() Fields {
	if f.Action == "" {
		f.Action = ActionUpdate
	}
	if f.Count < 1 {
		f.Count = 1
	}
	if f.Date == "" {
		f.Date = time.Now().UTC().Format("2006-01-02")
	}
	return f
}

// trailers builds the machine-readable body: the fixed keys of docs/06 in a
// fixed order, then any extra key sorted, so a message is reproducible.
func trailers(f Fields) []string {
	var out []string
	add := func(key, value string) {
		if value != "" {
			out = append(out, key+": "+value)
		}
	}
	add("Item", f.ItemID)
	add("Type", f.Type)
	switch {
	case f.PrevStatus != "" && f.Status != "" && f.PrevStatus != f.Status:
		add("Status", f.PrevStatus+" -> "+f.Status)
	default:
		add("Status", f.Status)
	}
	add("Project", f.ProjectKey)
	add("Board", f.Board)
	if f.Count > 1 {
		add("Items", strconv.Itoa(f.Count))
	}
	add("Tool", f.Tool)
	add("Agent", f.Agent)

	keys := make([]string, 0, len(f.Extra))
	for k := range f.Extra {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		add(k, f.Extra[k])
	}
	return out
}

// collapse folds a rendered subject into one line and trims it.
func collapse(in string) string {
	in = strings.ReplaceAll(in, "\r\n", " ")
	in = strings.ReplaceAll(in, "\n", " ")
	in = strings.ReplaceAll(in, "\r", " ")
	return strings.TrimSpace(strings.Join(strings.Fields(in), " "))
}

// truncate shortens a subject to n characters, ending it with a single ellipsis
// so that the reader can tell it was cut.
func truncate(in string, n int) string {
	runes := []rune(in)
	if len(runes) <= n {
		return in
	}
	return strings.TrimRight(string(runes[:n-1]), " ") + "…"
}

File: web/src/git/message.ts
/**
 * Commit message rendering for browser-only mode.
 *
 * The companion renders messages with Go's `text/template` inside
 * `internal/gitops`. The browser cannot run that code — git in the browser is
 * isomorphic-git, which shares nothing with the Go backend (ADR-006) — so the
 * one thing the two runtimes must agree on is the *format*, which is specified
 * in docs/06-git-sync.md §3.3 and implemented twice against the same cases.
 *
 * This module implements the documented subset: placeholder substitution and
 * the machine-readable trailers. It deliberately does not implement Go template
 * control flow (`{{if}}`, `{{range}}`, pipelines): a template that uses those is
 * a companion-only template, and `renderCommitMessage` says so rather than
 * rendering something different from what the companion would produce.
 */

/** What a write did, as `{{action}}` reports it. */
export type CommitAction = 'create' | 'update' | 'delete' | 'move' | 'comment';

/** The template context of docs/06 §3.3. */
export type CommitFields = {
  itemId?: string;
  title?: string;
  type?: string;
  status?: string;
  prevStatus?: string;
  projectKey?: string;
  board?: string;
  action?: CommitAction;
  /** Items covered by one commit; > 1 selects the bulk subject. */
  count?: number;
  user?: string;
  date?: string;
  /** `Tool:` trailer, for example `gintrack 0.4.1 (browser)`. */
  tool?: string;
  /** `Agent:` trailer, set when an agent made the change. */
  agent?: string;
};

/** A rendered message: the subject line and the trailer body. */
export type CommitMessage = { subject: string; body: string };

/** The shipped template (docs/06 §3.3). */
export const DEFAULT_COMMIT_TEMPLATE = 'pmngr: update {{.ItemID}} "{{.Title}}"';

/** Subject lines are truncated here; the full title survives in the body. */
export const SUBJECT_LIMIT = 72;

/** Both spellings of every placeholder, mapped onto one field. */
const PLACEHOLDERS: Record<string, keyof CommitFields> = {
  '.ItemID': 'itemId',
  id: 'itemId',
  '.Title': 'title',
  title: 'title',
  '.Type': 'type',
  type: 'type',
  '.Status': 'status',
  status: 'status',
  '.PrevStatus': 'prevStatus',
  prevStatus: 'prevStatus',
  '.ProjectKey': 'projectKey',
  project: 'projectKey',
  '.Board': 'board',
  board: 'board',
  '.Action': 'action',
  action: 'action',
  '.Count': 'count',
  count: 'count',
  '.User': 'user',
  user: 'user',
  '.Date': 'date',
  date: 'date',
};

/** A template that does not render the same way in both runtimes. */
export class CommitTemplateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CommitTemplateError';
  }
}

/**
 * Checks a template without rendering it, which is what the settings form does
 * before it saves. It throws `CommitTemplateError` with an actionable message.
 */
export function validateCommitTemplate(template: string): void {
  const source = template.trim() === '' ? DEFAULT_COMMIT_TEMPLATE : template;
  let depth = 0;
  for (let i = 0; i < source.length; i += 1) {
    if (source.startsWith('{{', i)) depth += 1;
    if (source.startsWith('}}', i)) depth -= 1;
    if (depth < 0) break;
  }
  if (depth !== 0) {
    throw new CommitTemplateError('the template has an unbalanced {{ … }}');
  }
  for (const match of source.matchAll(/\{\{\s*([^}]*?)\s*\}\}/g)) {
    const name = (match[1] ?? '').trim();
    if (!(name in PLACEHOLDERS)) {
      throw new CommitTemplateError(
        `unknown placeholder {{${name}}}: use ${Object.keys(PLACEHOLDERS)
          .filter((key) => !key.startsWith('.'))
          .join(', ')}`,
      );
    }
  }
}

/**
 * Renders a commit message. It mirrors `(*gitops.Template).Render`: the subject
 * is one line truncated to 72 characters, a truncated title is repeated in the
 * body, and the body carries the trailers in a fixed order.
 */
export function renderCommitMessage(template: string, fields: CommitFields): CommitMessage {
  const filled: CommitFields = {
    ...fields,
    action: fields.action ?? 'update',
    count: fields.count === undefined || fields.count < 1 ? 1 : fields.count,
    date: fields.date ?? new Date().toISOString().slice(0, 10),
  };

  let subject: string;
  if ((filled.count ?? 1) > 1 && (filled.itemId ?? '') === '') {
    // Several items in one commit: no id or title can name it (docs/06 §3.3).
    subject = `pmngr: ${filled.action ?? 'update'} ${String(filled.count)} items`;
  } else {
    validateCommitTemplate(template);
    const source = template.trim() === '' ? DEFAULT_COMMIT_TEMPLATE : template;
    subject = collapse(
      source.replace(/\{\{\s*([^}]*?)\s*\}\}/g, (_whole, name: string) => {
        const field = PLACEHOLDERS[name.trim()];
        const value = field === undefined ? undefined : filled[field];
        return value === undefined ? '' : String(value);
      }),
    );
  }
  if (subject === '') {
    throw new CommitTemplateError('the template rendered an empty subject');
  }

  const trailers = commitTrailers(filled);
  if (subject.length > SUBJECT_LIMIT) {
    trailers.unshift(`Title: ${filled.title ?? ''}`, '');
    subject = `${subject.slice(0, SUBJECT_LIMIT - 1).trimEnd()}…`;
  }
  return { subject, body: trailers.join('\n') };
}

/** The machine-readable trailers, in the fixed order of docs/06 §3.3. */
function commitTrailers(fields: CommitFields): string[] {
  const out: string[] = [];
  const add = (key: string, value: string | undefined) => {
    if (value !== undefined && value !== '') out.push(`${key}: ${value}`);
  };
  add('Item', fields.itemId);
  add('Type', fields.type);
  if (
    fields.prevStatus !== undefined &&
    fields.prevStatus !== '' &&
    fields.status !== undefined &&
    fields.status !== '' &&
    fields.prevStatus !== fields.status
  ) {
    add('Status', `${fields.prevStatus} -> ${fields.status}`);
  } else {
    add('Status', fields.status);
  }
  add('Project', fields.projectKey);
  add('Board', fields.board);
  if ((fields.count ?? 1) > 1) add('Items', String(fields.count));
  add('Tool', fields.tool);
  add('Agent', fields.agent);
  return out;
}

/** Folds a rendered subject into a single trimmed line. */
function collapse(value: string): string {
  return value.replace(/\s+/g, ' ').trim();
}
